convert/convert.go

```diff
@@ -0,0 +1,140 @@
+package convert
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/Clever/mongo-op-throttler/operation"
+	"gopkg.in/mgo.v2/bson"
+)
+
+// OplogBytesToOp converts the raw bytes for an oplog into a Mongo operation
+// as defined by operation.Op. There are two reasons we don't immediately write
+// the oplog entries to the database.
+// 1. Keeps the logic for understanding oplogs separate from the rest of the code
+// 2. Makes it easier to have a worker that takes in a file of operation.Ops instead
+// of the oplog
+func OplogBytesToOp(raw []byte) (*operation.Op, error) {
+	var bsonOp bson.M
+	if err := bson.Unmarshal(raw, &bsonOp); err != nil {
+		return nil, fmt.Errorf("Error parsing json: %s", err.Error())
+	}
+
+	return oplogEntryToOp(bsonOp)
+}
+
+// oplogEntryToOp converts from bson.M to operation.Op
+func oplogEntryToOp(oplogEntry bson.M) (*operation.Op, error) {
+	// Note that this has only been tested for the Mongo 2.4 format
+
+	// Based on the logic from the source code:
+	// https://github.com/mongodb/mongo/blob/v2.4/src/mongo/db/oplog.cpp#L791
+	opType, ok := oplogEntry["op"].(string)
+	if !ok {
+		return nil, fmt.Errorf("Missing op type")
+	}
+	namespace, ok := oplogEntry["ns"].(string)
+	if !ok {
+		return nil, fmt.Errorf("Missing namespace")
+	}
+
+	// Ignore changes to the system namespace. These are things like system.indexes
+	if strings.HasPrefix(namespace, "system.") {
+		return nil, nil
+	}
+
+	obj, ok := oplogEntry["o"].(bson.M)
+	if !ok {
+		return nil, fmt.Errorf("Missing object field")
+	}
+
+	switch opType {
+	case "i":
+		return convertToInsert(namespace, obj)
+	case "u":
+		return convertToUpdate(namespace, obj, oplogEntry)
+	case "d":
+		return convertToRemove(namespace, obj, oplogEntry)
+	default:
+		// It's theoretically possibly that is also 'c' or 'n', but we don't support them so let's error out
```

Comment:
c = cmd
n = noop

https://github.com/mongodb/mongo/blob/801d87f5c8d66d5f5a462c5e0daae67e6b848976/src/mongo/db/oplog.cpp#L147-L162

I will make this clearer in the code